### PR TITLE
Audit test and test comments

### LINKED ISSUE #735: Potential Replay Attack: Unprevented Edge Case
File: `pallets/msa/src/lib.rs`

Currently a signature registry is implemented to prevent replay attacks. The registry would delete all the signatures registered in a virtual bucket N using reset_virtual_bucket_if_needed(), when switching from the virtual bucket N to the next virtual bucket (i.e., the bucket N+1). Replay attack could happen if a signature is used near the end of the virtual bucket N, and the signature_expires_at falls within the bucket N+1.

## Scenario
1. Assume that MortalityWindowSize == 10, NumberOfBuckets == 10, and current_block == 9.
2. Alice signed a signature sig where signature_expires_at == 13 for Bob to transfer funds from Alice to Bob.
3. Bob successfully used sig when current_block == 9, and the chain registered sig.
4. Later when current_block == 10, the chain called reset_virtual_bucket_if_needed() and removed the previously registered sig.
5. Bob successfully used sig when current_block == 11, and the chain registered sig again.

### PR DESCRIPTION
**Pull request by shannonwells**
add clarifying comments, some rearranging and removal of duplicate tests, and demonstrate that a particular boundary case does not result in a replay

# Goal
The goal of this PR is 1) to add some comments on test cases as to what the point of each is, and b) to write up a test case proposed by the auditors to see if the test fails, and if so to fix it.

Closes #735 

# Discussion
1. is done
b. it doesn't fail

# Checklist
- [x] Tests added

### CODE REVIEW

File: pallets/msa/src/tests.rs [hunk 1]
```diff
@@ -2206,36 +2203,6 @@ pub fn clears_stale_signatures_after_mortality_limit() {
 	})
 }
 
-#[test]
-pub fn add_signature_replay_fails() {
```
Comment (shannonwells) [inline pallets/msa/src/tests.rs:2210]: These were duplicates of ones in replay_tests.

File: pallets/msa/src/replay_tests_2.rs [hunk 2]
```diff
@@ -0,0 +1,171 @@
+use crate::{self as pallet_msa, types::EMPTY_FUNCTION, AddKeyData, AddProvider, Config, Error};
```
Comment (shannonwells) [inline pallets/msa/src/replay_tests_2.rs:1]: Not sure how else to do proposed boundary checks which require different configs.  Maybe the filename could be renamed though.

File: pallets/msa/src/replay_tests.rs [hunk 3]
```diff
@@ -198,6 +210,11 @@ pub fn add_signature_replay_fails() {
 				Msa::register_signature(sig1, tc.mortality),
 				Error::<Test>::SignatureAlreadySubmitted,
 			);
+			run_to_block(tc.run_to + 1);
```
Comment (wilwade) [inline pallets/msa/src/replay_tests.rs:213]: Should this be just:
```suggestion
			run_to_block(tc.mortality);
```

File: pallets/msa/src/replay_tests.rs [hunk 4]
```diff
@@ -172,22 +172,34 @@ fn replaying_grant_delegation_fails() {
...
+			// This sets mortality at a boundary
+			TestCase { current: 11_149u64, mortality: 11_200u64, run_to: 11_199u64 },
+			// This does the same as above, but for a different bucket.
+			TestCase { current: 11_249u64, mortality: 11_300u64, run_to: 11_299u64 },
+			// This case sets current block at a boundary, and sets the mortality to the very last block before the boundary
 			TestCase { current: 1_100u64, mortality: 1_199u64, run_to: 1_198u64 },
-			TestCase { current: 1_102u64, mortality: 1_201u64, run_to: 1_200u64 },
-			TestCase { current: 1_099u64, mortality: 1_148u64, run_to: 1_101u64 },
+			// This case has current block right before a boundary, and sets expiration to the very last possible block
+			TestCase { current: 1_699u64, mortality: 1_798u64, run_to: 1_797u64 },
```
Comment (wilwade) [inline pallets/msa/src/replay_tests.rs:202]: All of the `run_to`s are `mortality - 1`. I guess that makes sense, but do we want `run_to` if that is the case?

File: pallets/msa/src/replay_tests.rs (hunk 3 above)
Comment (shannonwells) [reply pallets/msa/src/replay_tests.rs:213]: Yeah I think that's better.

**Review by enddynayn — APPROVED** [review]
